test/libsolidity/semanticTests/literals/ternary_operator_with_rationals.sol

```diff
@@ -0,0 +1,22 @@
+contract C {
+    function f(int a, int b) public returns (int) {
+        return a < b ? -1 : 1;
+    }
```

Comment:
I think this feature needs much more test coverage. Just looking at the code I can't really be sure that this will behave as expected in all the possible combinations. Our coverage of ternary operator with various types seems pretty sparse in general so we need to increase it.

Here are some suggestions of things I'd try to test:
- More sizes. E.g. `int8` vs `uint16`, etc. Also literals that are large enough to require these sizes.
- More types: `bytesNN`, booleans, strings. For example:
    - `x ? 1 : bytes1(1)`
    - `x ? "12345678" : bytes32(1)`
- `address` and `address payable`. The latter is convertible to the former and note that `address` literals exist (e.g. `0x1234567890123456789012345678901234567890` is of address type, not `uint160` or `bytes20`).
- Fixed point types. Do these work now?
    - `x ? 1.0 : 1.1`
    - `x ? 0 : 1.1`
- Are there any type combinations where both implicit conversions are possible so the result depends on whether we consider the true or false part first?
    - What about `x ? bs : bm` where `bs` is a storage bytes array and bm is memory bytes array? Will the result be one or the other type depending on the order?
- Array literals? Arrays vs slices? Slices are convertible to arrays under some conditions.

Also, it would be good to know which of these already worked and which ones are new.

Reply:
I added suggested test cases. Let me know if there is anything else that is missing.
Here are all test cases that were failing before and are successful after this PR changes:

**bytes1 v2 = cond ? bytes1("a") : "a";**
TypeError 1080: (142-166): True expression's type bytes1 does not match false expression's type string memory.

**bytes2 v3 = cond ? bytes2("a") : "a";**
TypeError 1080: (189-213): True expression's type bytes2 does not match false expression's type string memory.

**int8 v1 = cond ? -1 : 1;**
TypeError 1080: (82-95): True expression's type int8 does not match false expression's type uint8.

**int16 v2 = cond ? -32768 : 1;**
TypeError 1080: (117-134): True expression's type int16 does not match false expression's type uint8.

**int16 v5 = cond ? int16(1) : 1;**
TypeError 1080: (241-260): True expression's type int16 does not match false expression's type uint8.

**ufixed8x1 v7 = cond ? 1.0 : 1.1;**
TypeError 1080: (329-345): True expression's type uint8 does not match false expression's type ufixed8x1.**
TypeError 9574: (314-345): Type uint8 is not implicitly convertible to expected type ufixed8x1.

**ufixed8x1 v8 = cond ? 0 : 1.1;**
TypeError 1080: (371-385): True expression's type uint8 does not match false expression's type ufixed8x1.**
TypeError 9574: (356-385): Type uint8 is not implicitly convertible to expected type ufixed8x1.

**fixed8x1 v9 = cond ? -1.0 : -1.1;**
TypeError 1080: (410-428): True expression's type int8 does not match false expression's type fixed8x1.
TypeError 9574: (396-428): Type int8 is not implicitly convertible to expected type fixed8x1.

**fixed8x1 v10 = cond ? -1.1 : 1;**
TypeError 1080: (454-469): True expression's type fixed8x1 does not match false expression's type uint8.